`PwParser`: Refactor to primarily parse from XML and only parse from stdout what is necessary

sphuber opened the pull request:
Fixes #353

mbercx commented:
@sphuber: @t-reents and I have been working on getting the tests to pass in:

https://github.com/mbercx/aiida-quantumespresso/tree/fix/353/pw-parser-xml-only

Note that in https://github.com/mbercx/aiida-quantumespresso/commit/45b9183d4402e117b846e516dc9b49394cacf58a we decided that for the interrupted runs, we accept that we no longer parse as much information from the `stdout` as before. If we want to keep `stdout` parsing as a backup just in case the XML is not written properly for everything, then we have to keep maintaining the `stdout` parsing and I think this is exactly what we want to avoid here.

Now all tests are passing:

https://github.com/mbercx/aiida-quantumespresso/actions/runs/7279821777

Do you mind if I rebase this PR, add my changes here? I think we still need to do some final polishing of parts of the code, and a good ol' review, but we're close to finalising this!

sphuber commented:
Thanks for the help guys. Regarding the `stdout` parsing: I am afraid we unfortunately anyway have to keep it because all the error parsing comes from there. The XML reports nothing in terms of warnings or errors I believe, so all our error codes rely on the stdout. Or am I missing something?

t-reents commented:
No, that's correct. Moreover, something we were also discussing, a bunch of parameters that are present in the trajectory are only available in the stdout, e.g. magnetization in each step of a relaxation. The same is true for charges etc.

On the other hand, the final parameters are available in the XML as well.

sphuber commented:
Ideally, we would collect all data that we parse from the stdout that is not part of the XML and make a comprehensive list. We can then present this to the QE developers and ask whether (part of) that data can be made available in the XML. Especially parsing the various errors from XML would greatly robustify things I think. But this would be a lot of work and would only really bear fruits in quite a long time :D

mbercx committed:
♻️ `PwParser`: primarily parse from XML instead of `stdout`

As is extensively explained in a5637b3abafd88a770ec01687adff05f9ca07505, the parser of
the `pw.x` code of Quantum ESPRESSO was historically focused on parsing from the
`stdout`, since in the begin times there was no XML output to speak of. However, parsing
the `stdout` is often convoluted and fragile, breaking the parser with minor changes in
how the output is printed.

After updating the tests in a5637b3abafd88a770ec01687adff05f9ca07505, we update the
`PwParser` to rely on the XML for the majority of the output. To preserve backwards
compatibility, we still take data missing from the XML for supported Quantum ESPRESSO
versions from the `stdout` using the `backwards_compatibility_trajectory` and
`backwards_compatibility_parameters` methods.

We also add the units for several energy values that were previously missing in the
`output_parameters`.

Co-authored-by: Sebastiaan Huber <mail@sphuber.net>
Co-authored-by: Timo Reents <77727843+t-reents@users.noreply.github.com>
